### PR TITLE
419. Battleships in a Board

### PR DESCRIPTION
**Pull request by ErdemT09**
Resolves: https://leetcode.com/problems/battleships-in-a-board/

## Algorithm:
This is pretty simple:
Since the ships are only aligned vertically or horizontally, having a ship tile on the left tile or the tile above guarantees that we have counted that ship before, unless we are in the first column or row, then we always count the horizontal or the vertical, respectively.

### CODE REVIEW

File: src/main/java/algorithms/medium/BattleshipsInABoard.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    private final char BATTLESHIP = 'X';
+
+    public int countBattleships(char[][] board) {
+        int count = 0;
+        for (int i = 0; i < board.length; i++) {
+            for (int j = 0; j < board[0].length; j++) {
+
+                if (board[i][j] == BATTLESHIP && !hasBattleshipBefore(i, j, false, board)
+                        && !hasBattleshipBefore(i, j, true, board)) {
```
Comment (altayhunoglu) [inline src/main/java/algorithms/medium/BattleshipsInABoard.java:13]:
```suggestion
                if (board[i][j] == BATTLESHIP && !hasBattleshipBefore(i, j, board)
                        && !hasBattleshipBefore(i, j, board)) {
```

File: src/main/java/algorithms/medium/BattleshipsInABoard.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        return count;
+    }
+
+    private boolean hasBattleshipBefore(int i, int j, boolean checkLeft, char[][] board) {
+        if (checkLeft) {
+            return (j != 0 && board[i][j - 1] == BATTLESHIP);
+        } else checkAbove: {
+            return (i != 0 && board[i - 1][j] == BATTLESHIP);
+        }
+    }
```
Comment (altayhunoglu) [inline src/main/java/algorithms/medium/BattleshipsInABoard.java:28]:
```suggestion
     private boolean hasBattleshipBefore(int i, int j, char[][] board) {
            return (j != 0 && board[i][j - 1] == BATTLESHIP) ||   (i != 0 && board[i - 1][j] == BATTLESHIP);
     }
}
```

Comment (ErdemT09) [reply src/main/java/algorithms/medium/BattleshipsInABoard.java:28]: I wrote it out more explicitly, bu this is okay too.

**Review by altayhunoglu — APPROVED** [review]
Thank you for the valuable contribution, Erdem.

**Pull request merged**